File: baseline_modeling.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

# -------------------------------------------------------------
# 1. LOAD CLEANED DATA
# -------------------------------------------------------------
df = pd.read_csv("imdb_top_1000_cleaned_py.csv")

# Define feature sets
numeric_features = ["Released_Year", "Runtime_min", "Meta_score", "log_No_of_Votes", "log_Gross"]
categorical_features = ["Primary_Genre", "Certificate"]
target = "IMDB_Rating"

X = df[numeric_features + categorical_features]
y = df[target]

# -------------------------------------------------------------
# 2. TRAIN-TEST SPLIT & PREPROCESSING PIPELINE
# -------------------------------------------------------------
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.20, random_state=42
)

# Preprocessor: Scale numeric, One-Hot encode categorical
preprocessor = ColumnTransformer(
    transformers=[
        ("num", StandardScaler(), numeric_features),
        ("cat", OneHotEncoder(handle_unknown="ignore", sparse_output=False), categorical_features)
    ]
)

# Helper function to evaluate and return metrics
def evaluate_model(name, y_true, y_pred):
    rmse = np.sqrt(mean_squared_error(y_true, y_pred))
    mae = mean_absolute_error(y_true, y_pred)
    r2 = r2_score(y_true, y_pred)
    return {"Model": name, "RMSE": round(rmse, 4), "MAE": round(mae, 4), "R2": round(r2, 4)}

# -------------------------------------------------------------
# 3. MODEL 1: LINEAR REGRESSION (BASELINE)
# -------------------------------------------------------------
lr_pipeline = Pipeline(steps=[
    ("preprocessor", preprocessor),
    ("regressor", LinearRegression())
])

lr_pipeline.fit(X_train, y_train)
y_pred_lr = lr_pipeline.predict(X_test)
lr_metrics = evaluate_model("Linear Regression (Python)", y_test, y_pred_lr)

# -------------------------------------------------------------
# 4. MODEL 2: DECISION TREE REGRESSOR (WITH CV TUNING)
# -------------------------------------------------------------
dt_pipeline = Pipeline(steps=[
    ("preprocessor", preprocessor),
    ("regressor", DecisionTreeRegressor(random_state=42))
])

param_grid_dt = {
    "regressor__max_depth": [3, 5, 7, 10, 15],
    "regressor__min_samples_split": [2, 5, 10],
    "regressor__min_samples_leaf": [1, 2, 4]
}

dt_grid = GridSearchCV(
    dt_pipeline,
    param_grid_dt,
    cv=5,
    scoring="neg_root_mean_squared_error",
    n_jobs=-1
)
dt_grid.fit(X_train, y_train)

best_dt = dt_grid.best_estimator_
y_pred_dt = best_dt.predict(X_test)
dt_metrics = evaluate_model("Decision Tree Regressor (Python)", y_test, y_pred_dt)

# -------------------------------------------------------------
# 5. SUMMARY RESULTS & RESIDUAL VISUALIZATION
# -------------------------------------------------------------
results_df = pd.DataFrame([lr_metrics, dt_metrics])
print("\n=== Week 2 Python Model Evaluation on Test Set (80/20) ===")
print(results_df.to_string(index=False))
print(f"\nBest Decision Tree Parameters: {dt_grid.best_params_}")

# Residual Comparison Plot
plt.figure(figsize=(12, 5))

plt.subplot(1, 2, 1)
residuals_lr = y_test - y_pred_lr
sns.scatterplot(x=y_pred_lr, y=residuals_lr, alpha=0.7, color="navy")
plt.axhline(0, color="red", linestyle="--")
plt.title("Linear Regression: Residuals vs. Fitted")
plt.xlabel("Predicted IMDb Rating")
plt.ylabel("Residuals")

plt.subplot(1, 2, 2)
residuals_dt = y_test - y_pred_dt
sns.scatterplot(x=y_pred_dt, y=residuals_dt, alpha=0.7, color="darkgreen")
plt.axhline(0, color="red", linestyle="--")
plt.title("Tuned Decision Tree: Residuals vs. Fitted")
plt.xlabel("Predicted IMDb Rating")
plt.ylabel("Residuals")

plt.tight_layout()
plt.savefig("python_week2_residuals.png", dpi=300)
plt.close()

File: baseline_modeling.R
# Install packages if not already installed:
# install.packages(c("dplyr", "stringr", "caret", "rpart", "ggplot2", "e1071"))

library(dplyr)
library(stringr)
library(caret)
library(rpart)
library(ggplot2)

# -------------------------------------------------------------
# 1. LOAD & CLEAN DATA INLINE
# -------------------------------------------------------------
raw_df <- read.csv("/Users/sylvain/Sofia/MSCS3807 Data Modeling in Python and R/imdb-data-modeling/imdb_top_1000.csv", stringsAsFactors = FALSE)

# Preprocessing and Feature Engineering
df_clean <- raw_df %>%
  distinct(Series_Title, Released_Year, .keep_all = TRUE) %>%
  mutate(Released_Year = suppressWarnings(as.numeric(Released_Year))) %>%
  filter(!is.na(Released_Year)) %>%
  mutate(
    Runtime_min = as.numeric(str_replace(Runtime, " min", "")),
    Gross_clean = as.numeric(str_replace_all(Gross, ",", "")),
    Primary_Genre = str_trim(sapply(strsplit(as.character(Genre), ","), `[`, 1)),
    Certificate = ifelse(is.na(Certificate) | Certificate == "", "Unknown", Certificate)
  ) %>%
  mutate(
    Meta_score = ifelse(is.na(Meta_score), median(Meta_score, na.rm = TRUE), Meta_score),
    Gross_clean = ifelse(is.na(Gross_clean), median(Gross_clean, na.rm = TRUE), Gross_clean)
  ) %>%
  mutate(
    log_No_of_Votes = log1p(as.numeric(No_of_Votes)),
    log_Gross = log1p(Gross_clean)
  )

# Select modeling features and convert categoricals to factors
# Lump rare categories (< 15 occurrences) into "Other"
model_data <- df_clean %>%
  select(
    Released_Year, Runtime_min, Meta_score, log_No_of_Votes, log_Gross,
    Primary_Genre, Certificate, IMDB_Rating
  ) %>%
  mutate(
    Primary_Genre = forcats::fct_lump_min(Primary_Genre, min = 15, other_level = "Other"),
    Certificate = forcats::fct_lump_min(Certificate, min = 15, other_level = "Other")
  )

# -------------------------------------------------------------
# 2. TRAIN-TEST SPLIT & 5-FOLD CROSS-VALIDATION
# -------------------------------------------------------------
set.seed(42)
train_index <- createDataPartition(model_data$IMDB_Rating, p = 0.80, list = FALSE)
train_set <- model_data[train_index, ]
test_set  <- model_data[-train_index, ]

ctrl <- trainControl(method = "cv", number = 5)

calc_metrics <- function(name, y_true, y_pred) {
  rmse_val <- RMSE(y_pred, y_true)
  mae_val  <- MAE(y_pred, y_true)
  r2_val   <- R2(y_pred, y_true)
  data.frame(
    Model = name,
    RMSE = round(rmse_val, 4),
    MAE = round(mae_val, 4),
    R2 = round(r2_val, 4)
  )
}

# -------------------------------------------------------------
# 3. MODEL 1: LINEAR REGRESSION (BASELINE OLS)
# -------------------------------------------------------------
set.seed(42)
# Add "zv" (zero-variance removal) to preProcess
lr_model <- train(
  IMDB_Rating ~ .,
  data = train_set,
  method = "lm",
  preProcess = c("zv", "center", "scale"),
  trControl = ctrl
)

pred_lr <- predict(lr_model, newdata = test_set)
lr_res <- calc_metrics("Linear Regression (R caret)", test_set$IMDB_Rating, pred_lr)

# -------------------------------------------------------------
# 4. MODEL 2: DECISION TREE (rpart TUNING)
# -------------------------------------------------------------
set.seed(42)
dt_grid <- expand.grid(cp = seq(0.001, 0.05, by = 0.005))

dt_model <- train(
  IMDB_Rating ~ .,
  data = train_set,
  method = "rpart",
  tuneGrid = dt_grid,
  trControl = ctrl
)

pred_dt <- predict(dt_model, newdata = test_set)
dt_res <- calc_metrics("Decision Tree (R caret)", test_set$IMDB_Rating, pred_dt)

# -------------------------------------------------------------
# 5. SUMMARY RESULTS & RESIDUAL PLOT
# -------------------------------------------------------------
summary_results <- rbind(lr_res, dt_res)
cat("\n=== Week 2 R Model Evaluation on Test Set (80/20) ===\n")
print(summary_results)

test_set$Pred_LR <- pred_lr
test_set$Resid_LR <- test_set$IMDB_Rating - pred_lr

p_res <- ggplot(test_set, aes(x = Pred_LR, y = Resid_LR)) +
  geom_point(color = "navy", alpha = 0.6) +
  geom_hline(yintercept = 0, linetype = "dashed", color = "red") +
  theme_minimal() +
  labs(
    title = "Linear Regression (R): Residuals vs Fitted",
    x = "Predicted IMDb Rating",
    y = "Residuals"
  )

ggsave("r_week2_residuals.png", plot = p_res, width = 7, height = 5, dpi = 300)